toolchains/python/python.bzl

```diff
@@ -200,3 +200,184 @@ def nixpkgs_python_configure(
+
+
+def _nixpkgs_python_repository_impl(repository_ctx):
+    # 2. read generated json
+    python_modules = repository_ctx.read(repository_ctx.path(repository_ctx.attr.json_deps))
+
+    content = 'load("//:python_module.bzl", "python_module");'
+    for pkg_info in json.decode(python_modules):
+        pkg_name = pkg_info["name"]
+        pkg_store_path = pkg_info["store_path"]
+        deps = pkg_info["deps"]
+        pkg_link_path = "{}-link".format(pkg_name)
+        repository_ctx.symlink(pkg_store_path, pkg_link_path)
+
+        # Bazel chokes on files containing whitespaces, so we exclude them from
+        # the glob, hoping they are not important
+        content += """
+python_module(
+    name = "{name}",
+    store_path = "{link}",
+    files = glob(["{link}/**"], exclude=["{link}/**/* *"]),
+    deps = {deps},
+    visibility = ["//visibility:public"],
+)
+        """.format(name=pkg_name, link=pkg_link_path, deps=deps)
+
+    repository_ctx.file("BUILD.bazel", content)
+
+    # 3. generate BUILD.bazel file,
+    # ... _and_ the symlinks
+    # 4. Generate dummy WORKSPACE
+    # repository_ctx.file("WORKSPACE", "")
+
+    repository_ctx.file("python_module.bzl", """
+def _python_module_impl(ctx):
+    import_depsets = []
+    store = ctx.file.store_path
+    runfiles = ctx.runfiles(files = [store])
+
+    for dep in ctx.attr.deps:
+        runfiles = runfiles.merge(dep[DefaultInfo].data_runfiles)
+        import_depsets.append(dep[PyInfo].imports)
+
+    # HACK(danny): for some unforunate reason, short_path returns ../ when operating in external
+    # repositories. I don't know why. It breaks rules_python's assumptions though.
+    fixed_path = store.short_path[3:]
+    import_path = "/".join([ctx.workspace_name, store.short_path])
+
+    return [
+        DefaultInfo(
+            files = depset(ctx.files.files),
+            default_runfiles = ctx.runfiles(ctx.files.files, collect_default = True),
+        ),
+        PyInfo(
+            imports = depset(direct = [import_path], transitive = import_depsets),
+            transitive_sources = depset(transitive = [
+                dep[PyInfo].transitive_sources
+                for dep in ctx.attr.deps
+            ]),
+        ),
+    ]
+
+python_module = rule(
+    implementation = _python_module_impl,
+    attrs = {
+        "store_path": attr.label(
+            allow_single_file = True,
+            doc = "nix store path of python package",
+        ),
+        "files": attr.label_list(
+            allow_files = True,
+        ),
+        "deps": attr.label_list(
+            providers = [PyInfo],
+        ),
+    },
+    executable = False,
+    test = False,
+)
+""",
+    )
+
+    # 5. generate //:requirements.bzl for later import.
+    repository_ctx.file("requirements.bzl", """
+def requirement(package_name):
+    return "@{}//:{{}}".format(package_name)
+""".format(repository_ctx.name),
+    )
+
+    # TODO: make it lazy in the packages themselves ?
```

Comment:
I think that would require an extra indirection. I.e. generate a `.bzl` file that defines a macro that invokes `nixpkgs_package` for each package. I'm happy to leave this open as a future improvement.